java/core/src/java/org/apache/orc/OrcFile.java

```diff
@@ -642,4 +671,187 @@ public static Writer createWriter(Path path,
+    if (!reader.getSchema().equals(schema)) {
+      LOG.info("Can't merge {} because of different schemas {} vs {}",
+          path, reader.getSchema(), schema);
+      return false;
+    }
+    if (reader.getCompressionKind() != compression) {
+      LOG.info("Can't merge {} because of different compression {} vs {}",
+          path, reader.getCompressionKind(), compression);
+      return false;
+    }
+    if (compression != CompressionKind.NONE &&
+        reader.getCompressionSize() != compressionSize) {
+      LOG.info("Can't merge {} because of different compression sizes {} vs {}",
+          path, reader.getCompressionSize(), compressionSize);
+      return false;
+    }
+    if (reader.getFileVersion() != fileVersion) {
+      LOG.info("Can't merge {} because of different file versions {} vs {}",
+          path, reader.getFileVersion(), fileVersion);
+      return false;
+    }
+    if (reader.getWriterVersion() != writerVersion) {
+      LOG.info("Can't merge {} because of different writer versions {} vs {}",
+          path, reader.getFileVersion(), fileVersion);
+      return false;
+    }
+    if (reader.getRowIndexStride() != rowIndexStride) {
+      LOG.info("Can't merge {} because of different row index strides {} vs {}",
+          path, reader.getRowIndexStride(), rowIndexStride);
+      return false;
+    }
+    for(String key: reader.getMetadataKeys()) {
+      if (userMetadata.containsKey(key)) {
+        ByteBuffer currentValue = userMetadata.get(key);
+        ByteBuffer newValue = reader.getMetadataValue(key);
+        if (!newValue.equals(currentValue)) {
+          LOG.info("Can't merge {} because of different user metadata {}", path,
+              key);
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  static void mergeMetadata(Map<String,ByteBuffer> metadata,
+                            Reader reader) {
+    for(String key: reader.getMetadataKeys()) {
+      metadata.put(key, reader.getMetadataValue(key));
+    }
+  }
+
+  /**
+   * Merges multiple ORC files that all have the same schema to produce
+   * a single ORC file.
+   * The merge will reject files that aren't compatible with the merged file
+   * so the output list may be shorter than the input list.
+   * The stripes are copied as serialized byte buffers.
+   * The user metadata are merged and files that disagree on the value
+   * associated with a key will be rejected.
+   *
+   * @param outputPath the output file
+   * @param options the options for writing with although the options related
+   *                to the input files' encodings are overridden
+   * @param inputFiles the list of files to merge
+   * @return the list of files that were successfully merged
+   * @throws IOException
+   */
+  public static List<Path> mergeFiles(Path outputPath,
+                                      WriterOptions options,
+                                      Path... inputFiles) throws IOException {
```

Comment:
Really? It isn't clear to me that List<Path> is that much easier than a Path[] and the Path[] requires 1 fewer objects. But Ok.